---
title: "RNA-Sequencing - Differential Expression Analysis"
author: "UBEC"
date: "`r format(Sys.time(), '%Y/%m/%d')`"
output:
  html_document:
    code_download: true
    code_folding: hide
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: true
    toc_depth: '4'
params:
  config_file: "config.txt"
  project: ""
  readcount_file: ""
  metadata_file: ""
  custom_libs: ""
---
  
```{r setup, include=FALSE}


```

![](D:/ubec/DGE_report/ubec_logo.png)

Comments and general info about this script
```{r comment_code, warning=FALSE, message=FALSE}
#Rscript -e "rmarkdown::render(\\"DEG_template.Rmd\\", params = list(input =\\"\$PWD\\"), output_file =\\"\$PWD/${run_id}_dge.html\\")" 

#All preperation code, loading data, etc
###### START HERE ######
##### Step 1 ###########


# Note: Make sure that your design file contains only one column (main effect). The primary application of this DEG script are simple case/control designs.
#### Case/Control design (1 factor) ####### 
# If you specify a contrast, make sure that the reference level is the LAST variabel in the factor.
# E.x c("condition","case", "control")). This ensure that the fold-change is calculated relative to the reference level, not the other way around.

### Multi-factor design ######
### Combine variables across factor column in the design matrix to test for interaction effects ###
### See the DESEQ2 Vignette section on interactions for more information on how to combine factor levels.
### https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#interactions


# Do not use this script for complex designs with multiple factors etc.

# To DO
# 1. Set project name, w.dir
# 2. Set main effect (column in design)
# 3. Set design formula
# 4. Set design contrasts
# 5. Set annotation database
```

Main initialization. Check if required parameters are passed (or hardcoded when this script is run directly) and finally we set the basic parameters to their default values.  Load "DE_helper_functions.R" which contains the common functions used by this script and also loads the required libraries.

```{r initialization, warning=FALSE, message=FALSE}
project = NULL
metadata_file = NULL
readcount_file = NULL
annotation_data = NULL
group_replicates_by = NULL
batch = NULL
samples.in.comp=TRUE
export.plots=TRUE

perform.go=TRUE
perform.kegg=FALSE
plot.kegg=FALSE
#The comparisons witht the following for, can be hardcoded/overwritten here...

comparision <- list(c("condition", "A", "B"),
                    c("condition", "A", "C"),
                    c("condition", "B", "C")
)

if( (is.na(params$config_file) | params$config_file == '') && is.null(comparision) ) stop('Please specify the config file for the DGE analysis (--dge_config) or make sure it\'s specified' )

if( !is.null(params$config_file) && params$config_file != '') {
  source(params$config_file)
  #way to read in list of comparisions from config file...
  #scan reads data into vector/list using 'sep' as seperator and 'what' indicates data type, here text
  #gsub is used to  potential spaces surrounding the values
  #strsplit finally splits the character vector into substrings based on the ', '
  comparision <- strsplit(gsub(" ", "",scan(text=comparision, what="", sep="&"), fixed=TRUE), ",")
}

#knitr::opts_chunk$set
.libPaths(custom_libs)

source("DE_helper_functions.R")
```

```{r}
#setting 1 core is at least faster on local windows for DESeq method
setCores(1)

#Store workdir
w.dir=getwd()

#set design, 
#design <- ~batch+condition
design <- ~condition 

#handle special characters (such as '-')
comparision = map(comparision, correctIds)

#do we want to combine multiple count files (and add batch id to samples)
combined_run = F
addBatch = F

#Filter settings
alpha = 0.05  #fdr value
fc.cutoff = 1.5 #fold change cutoff, 1.5-2 depending on number of results...

#misc settings
#pre.filtering is used to filter out genes with a low nr of readcounts, mainly used for efficiency but not absolutely necessary as this is done by DESeq2 itself as well. There is no specific value to be used as cutoff. We use a cutoff of at least 5 reads total per row (gene) across all samples
pre.filtering = TRUE
write.normalized = TRUE #write the normalized count file
plot.mds = FALSE  #extra plots (addition to MDS, included for completeness but not part of the default output)
use_biomart = TRUE #What to use to retrieve gene names, biomart is the more up to date and complete source, otherwise bioconductor annotation packages will be used

#annotation database to use for linking gene names to ids
anno_db_data = NULL 

anno_db_data = init_annotation_db(use_biomart=T, species = species, annotation_data=annotation_data)

if(! is.null(anno_db_data)){
  anno_db = anno_db_data$anno_db
  gene_id_column = anno_db_data$gene_id_column
  gene_name_column = anno_db_data$gene_name_column
  kegg.genesets = anno_db_data$kegg.genesets
  go.seq.genome=anno_db_data$go.seq.genome
}

if(is.null(project)){
  stop('Please projectname for the DGE analysis (--project) or make sure it\'s specified (either in the config file or script itself)' )
}

if(! combined_run){
  #if specified in config, initialize to these values
  #manually specified overrides config
  if(!is.null(params$metadata_file) && !params$metadata_file == '') {
    metadata_file=params$metadata_file
  }
  #is still null, set based on projectid
  if(is.null(metadata_file)){ md_file=paste0(project, "_metadata.txt") }
  
  if(!is.null(params$readcount_file) && !params$readcount_file == ''){
    readcount_file <- params$readcount_file
  }
  if(is.null(readcount_file)){readcount_file <- paste0(project, "_exon_featureCounts.raw.txt")}
}
#check if condition is specified, add if not
if(comparision[[1]][1] != "condition"){
  x=function(l, x){l=c(x, l)}
  comparision=lapply(comparision, x, "condition")
}
```


# DE Report: `r project` 

![](D:/ubec/DGE_report/ubec_line.png)

# 1. Project Overview

## 1.1 Aims of the project

The aims of the project discussed in this report are highlighted in bold:
  
  **Data QC and sample comparison**
  
  **Differential expression analysis**
  
# 2. Files provided
  
  For the performed analyses, a number of files are produced, and figures are made to support the findings and their understanding. The type of files and figures that are provided are explained in detailed below.

-   CSV file with normalized data: raw counts corrected for sequencing differences between the samples. These normalized counts are at linear scale and are useful to get an idea of the expression changes across samples (note: not for within sample comparison of gene levels).

-   CSV files with differential expression lists per comparison. Both complete versions with no filtering based on p-value or fold change applied yet, so all genes of the dataset are included, as well as a filtered versions with cutoffs used for an adjusted p-value cutoff of `r alpha` and a log2 foldchange of `r round(log2(fc.cutoff))`. The differential expression results are optimized for an adjusted p-value cutoff of `r alpha`. The output in the tables contain the following:
  
  -   GENEID -- The ENSEMBL GENEID identifier

-   external_gene_name -- (Optional, only when annoated). The gene name/symbol for the GENEID

-   BaseMean -- Average of the normalized count values across all samples.

-   Log2FoldChange -- Effect size estimate log2(normalized counts group1/group2), positive values are upregulated in group1 cells. Each sample file is named "DE_group1vsgroup2" and should be used to infer the direction of the fold change.

-   lfcSE -- Uncertainty associated with the log2FC estimate.

-   P-value -- Wald test reflecting the probability that a fold change as strong as the observed one, would be seen under the situation described by the null hypothesis.

-   Padj -- P-value adjusted for multiple testing with Benjamini-Hochberg method.


-   Heatmap (HM): is useful to visualize a list of multiple genes across samples (e.g. the list of differentially expressed genes of a specific comparison), instead of looking at the expression pattern of individual genes. The represented expression values are normalized transformed counts. The genes (rows) are clustered based on similarities in their expression pattern across the samples. The normalized transformed counts are shown, scaled per gene (representing the deviation from the average across samples).

-   The gene plots show the normalized counts for the top10 significant genes (Padj\<`r alpha`) with highest fold change (gene count on linear and log10 scale), to illustrate the expression across the samples. The adjusted P-value and log2 fold change for each gene are indicated underneath each graph.

# 3. Methods

## 3.1 Sample similarity and 2D representation

The overall similarity between samples is assessed by calculating the Euclidean distances and represented in a heatmap. Principal component analysis helps to visualize the underlying structure based on the variation in the data. For both approaches, the data is transformed using Variance Stabilizing Transformation to ensure roughly equal contribution from all genes.

## 3.2 Differential expression (using general linear models)

For the comparisons which combines samples, DESeq2 was used, an R-package that provides methods for differential expression testing based on linear models. DESeq2 models the raw counts, using normalization factors to account for differences in library depth. Then, it estimates the gene-wise dispersions (variability) and shrinks these estimates to generate more accurate estimates of dispersion to model the counts. Finally, DESeq2 will fit the negative binomial model and perform hypothesis testing using the Wald test for each gene. A Wald test statistic is computed based on the estimated fold change along with a probability (the p-value) that a test statistic at least as extreme as the observed value were selected at random. We essentially want to determine whether the mean expression levels of different sample groups are significantly different. The log2 fold changes are adjusted using Bayesian shrinkage (sh_log2FC). The log2 fold changes of genes with little information or high dispersion (variation) are adjusted toward more likely lower log2 fold change estimates. This fold change value is recommended to use for ordering or when using a fold change cutoff for downstream analysis of the gene list. This fold change is used for representation (MA-plots and indicated heatmap).

The statistics of the differential expression analyses are represented in MA-plots and volcano plots. The expression profiles of the list of differentially expressed genes are shown in heatmaps (selection of DE genes based on adjusted P-value and fold change). The top10 differentially regulated genes (higher fold change) are individually represented across samples.

# 4. Results

Read in readcount data and metadata files
```{r read_data, warning=FALSE, message=FALSE}
if(combined_run) {
  #read in collection of files
  readcount_files<- list.files(getwd(), pattern = "_featureCounts.raw.txt", full.names = T)
  raw_counts<- map(readcount_files, read_in_feature_counts_custom)
  
  #fix column name old files if necessary
  #colnames(raw_counts[[1]])[1] = "Geneid"
  
  rawcounts = Reduce(merge_counts, raw_counts) 
  row.names(rawcounts) <- rawcounts$Geneid
  rawcounts = rawcounts[, -1]
  
  #create metadata  
  metadata_files<- list.files(getwd(), pattern = "_MetaData.txt", full.names = T)
  if(length(metadata_files) ==0) metadata_files<- list.files(getwd(), pattern = "_metadata.txt", full.names = T)
  
  #cneed to heck if metadatafiles can be combined, nr columns and column names should match
  
  md <- map(metadata_files, readMetadataFile)
  md = Reduce(rbind, md) 
  rownames(md) = correctIds(md$sampleid)
  
  #add batch information to metadata file
  if(addBatch ==T){
    
    #construct dataframe for batch info
    #set size
    samplecount=0
    for(i in 1:length(raw_counts)){
      samplecount = samplecount+ncol(raw_counts[[i]])-1
    }
    #create initial dataframe
    batch_map <- data.frame(sample=rep("", samplecount), batch=rep("", samplecount), stringsAsFactors=FALSE)
    
    #start filling
    row=1
    for(i in 1:length(raw_counts)){
      tmp = raw_counts[[i]]
      
      for(j in 2:ncol(tmp)){
        batch_map[row, ] = list(colnames(tmp)[j], i)
        row = row+1
      }
    }
    rownames(batch_map) = batch_map$sample
    
    #and add to metadata
    md$batch=-1
    for(i in 1:ncol(rawcounts)){
      sample=colnames(rawcounts)[i]
      md[sample, ]$batch = batch_map[sample, ]$batch
    }
  }
} else {
  #otherwise simply read in single files
  
  #Read in MetaData
  #md <- readMetadataFile(paste0(project, "_metadata.txt"))
  md <- readMetadataFile(metadata_file)
  #correct md, first set rownames
  rownames(md) = correctIds(md$sampleid)
  
  #Read raw readcounts
  #rawcounts=read_in_feature_counts_custom(readcount_file)
  
  rawcounts=readCountsFile(readcount_file)
  #store gene lengths
  #genelengths=data.frame(rawcounts$length)
  genelengths=data.frame(rawcounts$Length)
  rownames(genelengths)=rownames(rawcounts)
  colnames(genelengths) =c("length")
  
  #and remove unneeded columns
  rawcounts = removeColumnsFromCountTable(rawcounts) 
  #rawcounts = removeColumnsFromCountTable(rawcounts, remove.columns = c("chr", "start", "end", "strand", "length"))
  #only keep counts for samples actually in metadata
  rawcounts = rawcounts[, rownames(md)]
  
  #colnames(rawcounts)[1] = "Geneid"
  #row.names(rawcounts) <- rawcounts$Geneid
}

md$condition_org = md$condition
#correct conditions where necessary (a '-' causes issues in comparisons)
md$condition = correctIds(md$condition_org)

#determine minimal group size to set cutoff for counts
#min_groupsize=getMinimalGroupSize(md)
min_groupsize = 2  #above too stringent for larger groups especially
```

Start with DE analysis processing, create a "DE" subfolder to store seperate plots and files.

```{r de_analysis_prep, warning=FALSE, message=FALSE}
#Change to directory where results will be written
dir.create(file.path(w.dir, "DE"), showWarnings = FALSE)
setwd(paste0(w.dir, "/DE"))
current_wd = getwd()

if(export.plots){
  dir.create(file.path(w.dir, "plots"), showWarnings = FALSE)
}

dds <- DESeqDataSetFromMatrix(countData = rawcounts[, rownames(md)],
                              colData = md,
                              design = design)

if(!is.null(group_replicates_by)){
  #dds = collapseReplicates(dds, dds$subjectid)
  dds = collapseReplicates(dds, dds[[group_replicates_by]])
}


#Pre-filter low count values
if (pre.filtering) {
  dds <- estimateSizeFactors(dds) #not needed but doesnt harm
  #idx1 <- rowSums(counts(dds, normalized = T) >= 5) >= 2
  #dds <- DESeq(dds, parallel = T)
  idx <- rowSums(counts(dds, normalized = T) >= 5) >= 2
  dds <- dds[idx,]
  
}
dds <- DESeq(dds, parallel = T, test = "Wald")
if (write.normalized) {
  #dds <- estimateSizeFactors(dds) #not needed but doesnt harm
  counts.norm <- counts(dds, normalized=TRUE)  
  write.csv(data.frame("GENEID"=rownames(counts.norm),counts.norm), file = paste0(project,"_counts_normalized",".csv"), quote = F, row.names = F)
}
```

## 4.1 Sample comparison

To evaluate the relatedness between samples, the Euclidean distances between the samples and principal components in the dataset are determined and represented below (Figure 1 and Figure 2).  

```{r generate_heatmap, fig.height=14, fig.width=16}
sample_hm=generateHeatmap(dds=dds, title=paste0("Heatmap ",project))
if(export.plots==TRUE){
  save_pdf(sample_hm, filename=paste0("plots/Heatmap ",project, ".pdf"))
  save_svg(sample_hm, filename=paste0("plots/Heatmap ",project, ".svg"))
}
```

*Figure 1 Heatmap showing the correlation between the transcriptional profiles of all samples.*
  
```{r generate_pca, fig.height=14, fig.width=16}
vsd <- vst(dds, blind=TRUE)
#rld <- rlog(dds, blind = TRUE) #vst is a faster version rld and suitable as well 
pcadata<- plotPCA(vsd, intgroup = colnames(md),  returnData=T)
#addBatch=T
if(!is.null(batch)){
  #generatePCARepel(pcadata,"condition", "batch", project)
  pcaplot=generatePCARepel(pcadata, batch, "condition", project)
} else {
  pcaplot=generatePCARepel(pcadata=pcadata, color="condition", title=project)
}

if(export.plots==TRUE){
  save_pdf(pcaplot, filename=paste0("plots/PCA_",project, ".pdf"))
  save_svg(pcaplot, filename=paste0("plots/PCA_",project, ".svg"))
}
```

*Figure 2 Principal component plots of all samples reveal the underlying structure in the expression data of all samples in 2D.*
  
## 4.2 Differential gene expression analysis - Comparisons {.tabset .tabset-pills}
  
Here follows an overview of the DE results for each specified comparison...

```{r datatable_init, echo=FALSE,include = FALSE}
# You need this code to conduct the magic dependences attaching... Needed in order to get table rendered for some reason.
#https://github.com/rstudio/DT/issues/67
DT::datatable(matrix())
#htmltools::tagList(datatable(matrix()))
```


```{r de_analysis_new, cache=FALSE, warning=FALSE, message=FALSE, echo=FALSE, results="asis", fig.height=10, fig.width=12}
comparison_counter=1
figure_counter=3
comp=comparision[[1]]

#setup column info for the dge gene heatmap later on...
sampleInfo=as.data.frame(colData(dds))

for (comp in comparision) 
{
  sub_dir=paste0(comp[2], "_vs_", comp[3])
  dir.create(file.path(current_wd, sub_dir), showWarnings = FALSE)
  setwd(file.path(current_wd, sub_dir))
  
  contrast =  c(comp[1],comp[2], comp[3])
  results_contrast = getDataForContrast(
    dds = dds,
    contrast=contrast,
    anno_db = anno_db,
    fdrval = alpha,
    fcval=fc.cutoff)
  
  #added to only show samples actually in current comparison in heatmap
  sampleInfoComparison=sampleInfo
  if(samples.in.comp==TRUE){
    sampleInfoComparison = sampleInfo[sampleInfo$condition==comp[2] | sampleInfo$condition==comp[3], ] %>% dplyr::select(all_of(c("condition")))
  }
  
  # read in DE genes
  de.genes = read.csv(file = paste0(contrast[1],"_",contrast[2],"_vs_",contrast[3],"_padj_",alpha,".csv"), header=T)
  row.names(de.genes) = de.genes[, "GENEID"]
  de.genes = subset(de.genes, select=-c(Row.names))
  
  #one tabset per comparison
  cat("  \n### ", comp[2], " vs ", comp[3], " {.tabset .tabset-pills} \n", sep = "")
  
  cat("The gene expression differences between ", comp[2], " and ", comp[3], " samples was assessed. In total ", nrow(results_contrast[results_contrast$filter!=0,]), "genes were found significantly differentially regulated with an adjusted P-value <", alpha, " and a fold change exceeding ", fc.cutoff, " (|log2FC|>", round(log2(fc.cutoff), digits=2), "). \n \n" )
  
  #display table of results
  cat("  **Results:** \n\n")
  
  dt = datatable_custom(data=de.genes, row.names = F, round.cols = c("log2FoldChange", "lfcSE", "stat"), signif.cols = c("pvalue", "padj"))
  cat(knitr::knit_print(dt))
  cat(' \n \n')
  
  #Heatmap creation
  cat("  \n#### Heatmap DE genes\n", sep = "")
  
  cat(" This heatmap shows at most the top 50 of most differentially expressed gene between the conditions.\n")
  
  maxfc=as.integer(max(abs(min(results_contrast$log2FoldChange)), max(results_contrast$log2FoldChange)))
  hm_count=min(50, nrow(de.genes) )
  if(hm_count>0){
    rowLabels = paste0(de.genes$GENEID[1:hm_count], " | ", de.genes$external_gene_name[1:hm_count])
    comp_hm=pheatmap(assay(vsd)[de.genes$GENEID[1:hm_count], rownames(sampleInfoComparison)], labels_row=rowLabels, annotation_col=sampleInfoComparison, scale='row') 
    
    if(export.plots==TRUE){
      save_pdf(comp_hm, filename=paste0("../../plots/DE_Heatmap_",project, "_", comp[2], "-", comp[3],".pdf"))
      save_svg(comp_hm, filename=paste0("../../plots/DE_Heatmap_",project, "_", comp[2], "-", comp[3],".svg"))
    }
    
    figure_counter = figure_counter+1
  }
  cat(' \n \n')
  cat("  *Figure ",figure_counter, " Heatmap of top 50 (max) differential expressed genes based on p-value.*")
  cat(' \n \n')
  
  #tab for volcano+ma plots
  cat("  \n#### Volcano and MA-plot \n", sep = "")
  
  cat("  **Volcano plot** \n\n")
  cat("Volcano plots are commonly used to display the results of RNA-seq or other omics experiments. A volcano plot is a type of scatterplot that shows statistical significance (P value) versus magnitude of change (fold change). It enables quick visual identification of genes with large fold changes that are also statistically significant. These may be the most biologically significant genes. In a volcano plot, the most upregulated genes are towards the right, the most downregulated genes are towards the left, and the most statistically significant genes are towards the top.")
  cat(' \n \n')
  
  volcanoplot=plotVolcano(results_contrast=results_contrast, contrast= c(comp[1],comp[2], comp[3]), fcval=fc.cutoff, fc.max=maxfc, fdrval=alpha)
  cat(' \n \n')
  cat("  *Figure ",figure_counter, " Volcano-plot (scatter plotthat shows statistical significance (P value) versus magnitude of change (log2 fold change)).*")
  figure_counter = figure_counter+1
  cat(' \n \n')
  
  cat("  **MA-plot** \n \n")
  cat("This displays the global view of the relationship between the expression change of conditions (log ratios, M), the average expression strength of the genes (average mean, A), and the ability of the algorithm to detect differential gene expression. The genes that passed the significance threshold (adjusted p-value < 0.01) are colored with the genes being down-regulated being colored in blue while the genes being up-regulated are colored in red and each dot representing a gene.")
  cat(' \n \n')
  
  
  maplot=plotMA(results_contrast=results_contrast, contrast= c(comp[1],comp[2], comp[3]), fc.max=maxfc, fdrval=alpha)
  
  if(export.plots==TRUE){
      save_pdf(volcanoplot, filename=paste0("../../plots/DE_Volcano_",project, "_", comp[2], "-", comp[3],".pdf"))
      save_svg(volcanoplot, filename=paste0("../../plots/DE_Volcano_",project, "_", comp[2], "-", comp[3],".svg"))
      save_pdf(maplot, filename=paste0("../../plots/DE_MA_",project, "_", comp[2], "-", comp[3],".pdf"))
      save_svg(maplot, filename=paste0("../../plots/DE_MA_",project, "_", comp[2], "-", comp[3],".svg"))
    }
  cat(' \n \n')
  cat("  *Figure ",figure_counter, " MA-Plot (scatter plot of log2 fold changes (on the y-axis) versus the mean of normalized counts (on the x-axis)*")
  cat(' \n \n')
  figure_counter = figure_counter+1
  
  if(perform.kegg){
    ### KEGG pathway analysis
    cat("  \n#### KEGG pathway analysis", " {.tabset .tabset-pills} \n", sep = "")
    #    cat(' \n \n')
    cat('  **KEGG Pathway analysis** \n \n')
    cat(" Another popular annotation database is the Kyoto Encyclopedia of Genes and Genomes (KEGG). This is a curated database of molecular pathways and disease signatures. This KEGG analysis is done using the GAGE R-module. \n \n")
    #cat(' \n \n')
    cat("KEGG Pathway database: https://www.genome.jp/kegg/pathway.html \n \n")
    cat("GAGE R-package: https://bioconductor.org/packages/release/bioc/html/gage.html \n \n")
    
    #construct df for kegg pathway analysis converting ensembl ids to entrez ids (required for kegg pathway analysis) 
    deseq.fc = results_contrast$log2FoldChange
    entrezIds = ensemblToEntrezIds(ensembl.ids = row.names(results_contrast), biomart.db = anno_db)
    names(deseq.fc) = entrezIds$entrezgene_id
    
    #perform pathway analysis
    fc.kegg.anydir <- gage(deseq.fc, gsets = kegg.genesets, same.dir = F)
    fc.kegg.samedir <- gage(deseq.fc, gsets = kegg.genesets, same.dir = T)
    #depending on same.dir parameter plot top 2 from each direction or top 3 in general
    path.ids.a <- rownames(fc.kegg.anydir$greater)[c(1:3)]
    path.ids.l <- rownames(fc.kegg.samedir$less)[c(1:3)]
    path.ids.g <- rownames(fc.kegg.samedir$greater)[c(1:3)]
    
    cat("  \n##### Up regulated \n", sep = "")
    
    #print(knitr::kable(fc.kegg.samedir$greater[c(1:10),]))
    dt = datatable_custom(data=fc.kegg.samedir$greater, row.names = T, round.cols = c("stat.mean"), signif.cols = c("p.geomean", "p.val", "q.val", "exp1"))
    cat(knitr::knit_print(dt))
    
    cat(' \n \n')
    
    if(plot.kegg){
      for(id in substr(c(path.ids.g), 1, 8)){
        plot_pathview(genelist = deseq.fc, pathwayid = id)
        cat(' \n \n')
      }
    }
    
    cat("  \n##### Down regulated\n", sep = "")
    
    #print(knitr::kable(fc.kegg.samedir$less[c(1:10),]))
    dt = datatable_custom(data=fc.kegg.samedir$less, row.names = T, round.cols = c("stat.mean"), signif.cols = c("p.geomean", "p.val", "q.val", "exp1"))
    cat(knitr::knit_print(dt))
    
    cat(' \n \n')
    
    if(plot.kegg){
      for(id in substr(c(path.ids.l), 1, 8)){
        plot_pathview(genelist = deseq.fc, pathwayid = id)
        cat(' \n \n')
      }
    }
    cat("  \n##### Any regulated\n", sep = "")
    cat(' \n \n')
    
    #print(knitr::kable(fc.kegg.anydir$greater[c(1:10),]))
    dt = datatable_custom(data=fc.kegg.anydir$greater, row.names = T, round.cols = c("stat.mean"), signif.cols = c("p.geomean", "p.val", "q.val", "exp1"))
    cat(knitr::knit_print(dt))
    
    cat(' \n \n')
    
    if(plot.kegg){
      for(id in substr(c(path.ids.a), 1, 8)){
        plot_pathview(genelist = deseq.fc, pathwayid = id)
        cat(' \n \n')
      }    
    }
  }
  
  if(perform.go){
    #tab for GO
    cat("  \n#### GO-term analysis  {.tabset .tabset-pills}\n", sep = "")
    cat('  **GO-term analysis** \n \n')
    cat("One simple, but extremely widely used, systems biology technique for highlighting biological processes is gene category over-representation analysis. In order to perform this analysis, genes are grouped into categories by some common biological property and then tested to find categories that are over represented amongst the differentially expressed genes.")
    cat(' \n \n')
    
    cat("For this we use the goseq R-package: \n https://bioconductor.org/packages/release/bioc/html/goseq.html")
    cat(' \n \n')
    
    ### GO-term analysis
    #get de genes based on earlier filter (FC and pvalue)
    de.genes=rownames(results_contrast[results_contrast$filter!=0, ])
    all.genes=rownames(results_contrast)
    
    #version with lengths manually retrieveed from count file...
    gene.lengths = genelengths[rownames(genelengths) %in% all.genes,]
    
    #bias.data null would use genome inof
    all.go.results=go.analysis.goseq(de.genes = de.genes, all.genes = all.genes, genome = go.seq.genome, geneid.source = "ensGene", bias.data=gene.lengths )
    
    cat("  \n*Figure ",figure_counter, " GO-analysis*\n")
    figure_counter = figure_counter+1
    
    #tab for each GO category
    cat("  \n##### Biological Process\n", sep = "")
    
    go.data=all.go.results[all.go.results$ontology=="BP", ]
    enriched.go = go.data[go.data$over_represented_pvalue<0.05,]
    dt = datatable_custom(data=enriched.go[c(1:min(250, length(enriched.go))), ], row.names = F, signif.cols = c("over_represented_pvalue", "under_represented_pvalue"))
    cat(knitr::knit_print(dt))
    cat(' \n \n')
    
    enriched.file = paste0(contrast[1],"_",contrast[2],"_vs_",contrast[3],"_BP_enriched.txt")
    process.go.results(go.data=all.go.results[all.go.results$ontology=="BP", ], p.value=0.05, enriched.file = enriched.file)
    
    cat(' \n \n')
    
    cat("  \n*Figure ",figure_counter, " GO-analysis*\n")
    figure_counter = figure_counter+1
    
    cat("  \n##### Cellular Component\n", sep = "")
    
    go.data=all.go.results[all.go.results$ontology=="CC", ]
    enriched.go = go.data[go.data$over_represented_pvalue<0.05,]
    dt = datatable_custom(data=enriched.go[c(1:min(250, length(enriched.go))), ], row.names = F, signif.cols = c("over_represented_pvalue", "under_represented_pvalue"))
    cat(knitr::knit_print(dt))
    cat(' \n \n')
    
    enriched.file = paste0(contrast[1],"_",contrast[2],"_vs_",contrast[3],"_CC_enriched.txt")
    process.go.results(go.data=all.go.results[all.go.results$ontology=="CC", ], p.value=0.05, enriched.file = enriched.file)
    
    cat(' \n \n')
    
    cat("  \n##### Molecular Function\n", sep = "")
    
    go.data=all.go.results[all.go.results$ontology=="MF", ]
    enriched.go = go.data[go.data$over_represented_pvalue<0.05,]
    dt = datatable_custom(data=enriched.go[c(1:min(250, length(enriched.go))), ], row.names = F, signif.cols = c("over_represented_pvalue", "under_represented_pvalue"))
    cat(knitr::knit_print(dt))
    cat(' \n \n')
    
    enriched.file = paste0(contrast[1],"_",contrast[2],"_vs_",contrast[3],"_MF_enriched.txt")
    process.go.results(go.data=all.go.results[all.go.results$ontology=="MF", ], p.value=0.05, enriched.file = enriched.file)
    
    cat(' \n \n')
    cat("  \n*Figure ",figure_counter, " GO-analysis*\n")
    figure_counter = figure_counter+1
  }
  
  setwd(current_wd)
  comparison_counter = comparison_counter+1
  
  cat("  \n")
}
```

```{r exit_code, include=FALSE}
#needed to get buttons, filtering, etc to work ()
knitr::knit_print(datatable_custom(matrix(), row.names=F))

knitr::knit_exit()
```

## 4.3 Provided files

An overview of the various files supplied

```{r print_file_structure, echo=FALSE, warning=FALSE, message=FALSE, results="asis"}
cat("  \n### 4.3.1 DE Report\n")
cat("  \nThis file...  \n")

cat("  \n### 4.3.2 General plots and data\n")

cat("  *Heatmap (Figure 1):*")
cat(paste0("  \n./DE/HM_",project,".pdf  \n"))
cat("\n")

cat("  *PCA Plot (Figure 2):*")
cat(paste0("  \n./DE/PCA_",project,".pdf  \n"))
cat("\n")

cat("  *Normalized counts CSV file:*")
cat(paste0("  \n./DE/", project, "_counts_normalized.csv  \n"))

comparison_counter=3
figure_counter=3

for (comp in comparision) {
  contrast =  c(comp[1],comp[2], comp[3])
  cat("  \n### 4.3.", comparison_counter, " Comparison ", comp[2], " vs ", comp[3], " plots and data\n", sep = "")
  
  
  cat("  *QC-plots (Volcano and MA, figure ", figure_counter, " and ", (figure_counter+1), "):*", sep="")
  cat(paste0("  \n./DE/", comp[2], "_vs_", comp[3], "/QC_", contrast[1],"_",contrast[2],"_versus_",contrast[3],".pdf  \n"))
  
  cat("\n")
  cat("  *Differential expression CSV list (unfiltered):*")
  cat(paste0("  \n./DE/", comp[2], "_vs_", comp[3], "/condition_", contrast[1],"_",contrast[2],"_vs_",contrast[3],".csv  \n"))
  
  cat("\n")
  cat("  *Differential expression CSV list (p-adj & foldchange filtered):*")
  cat(paste0("  \n./DE/", comp[2], "_vs_", comp[3], "/condition_", contrast[1],"_",contrast[2],"_vs_",contrast[3],"_padj_",alpha,".csv  \n"))
  
  comparison_counter = comparison_counter+1
  figure_counter=figure_counter+2
}
```
