# RNASeq_DE_Analysis
R-script pipeline for DE analysis of RNAseq data as generated by USEQ

These are the scripts used by the UBEC to perform the basic Differential Expression analysis on RNASeq data. 

The main workbook is the Rmd file, containing the various steps in pre- and post-processing and which is used as the basis to generate the final HTML report.

The DE_helper_functions.R file contains most of the functions used by the Rmd file and is in a seperate file so it might also be used for other projects.

## Usage

First of all make sure all the necessary libraries are installed, these are also listed in the DE_helper_functions.R file.

### Basic package installation

By default the differential expression analysis contains a the heatmap and PCA plot of the samples and the results of the DE analysis using DESeq2 including the heatmaps, volcano plots, MA-plots and tables of the differntially expressed genes.

Install the required packages for the basis DE analysis:

```
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  
  BiocManager::install("BiocParallel")
  BiocManager::install("biomaRt")
  BiocManager::install("DESeq2")
  BiocManager::install("clusterProfiler")
    
  install.packages("tidyverse")	#purrr, ggplpot2, dplyr, tidyr, readr, tibble, stringr, forcats. https://www.tidyverse.org/packages/
  install.packages("RColorBrewer")
  install.packages("pheatmap")
  install.packages("ggplot2")
  install.packages("ggrepel")
  install.packages("DT")
  install.packages("ggpubr")
  install.packages('rmarkdown')
  install.packages('plotly')
```

### Annotation packages

While .Rdata files which can be used in the basic DE analysis are supplied, when not used the following annotation packages are required (depending on the species of interest). They are really required when one wants to run the extended GO-Term and pathway analysis as described next.

```
#depends on target species
#human and mouse, most commmon species
BiocManager::install("org.Mm.eg.db")
BiocManager::install("org.Hs.eg.db")
#dog
#BiocManager::install("org.Cf.eg.db")
```

### Extended package installation

Currently in development is also the inclusion of GO-Term and Pathway analysis modules. For these GOSeq and Gage are being used, while this is in development and not yet used one can also try to run this part and for that the following libraries are needed.

Install the required packages for the extended (pathway and go-term) DE analysis:

```
  BiocManager::install("gage")
  BiocManager::install("gageData")
  BiocManager::install("pathview")
  
  BiocManager::install("goseq")
  #already installed by gage and/or goseq, included for completeness
  #BiocManager::install("GO.db")
```
